httpbin/middleware.go

```diff
@@ -163,22 +164,27 @@ type Observer func(result Result)
 
 // StdLogObserver creates an Observer that will log each request in structured
 // format using the given stdlib logger
-func StdLogObserver(l *log.Logger) Observer {
-	const (
-		logFmt  = "time=%q status=%d method=%q uri=%q size_bytes=%d duration_ms=%0.02f user_agent=%q client_ip=%s"
-		dateFmt = "2006-01-02T15:04:05.9999"
-	)
+func StdLogObserver(l *slog.Logger) Observer {
+	const dateFmt = "2006-01-02T15:04:05.9999"
+	var logLevel = slog.LevelInfo
+
 	return func(result Result) {
-		l.Printf(
-			logFmt,
-			time.Now().Format(dateFmt),
-			result.Status,
-			result.Method,
-			result.URI,
-			result.Size,
-			result.Duration.Seconds()*1e3, // https://github.com/golang/go/issues/5491#issuecomment-66079585
-			result.UserAgent,
-			result.ClientIP,
+		if result.Status >= 500 {
+			logLevel = slog.LevelError
+		} else if result.Status >= 400 && result.Status < 500 {
+			logLevel = slog.LevelWarn
+		}
```

Comment:
This isn't safe, we're modifying a `logLevel` var defined outside this function on every invocation.  I _think_ that means a single 5xx response followed by a 2xx response will be logged at error level.

I'd probably do something like this instead:

```go
logLevel := slog.InfoLevel
if result.Status >= 500 {
    logLevel = slog.LevelError
} else if result.Status >= 400 {
    logLevel = slog.LevelWarn
}
```

---

Yep, confirmed locally with a `/status/500` then `/status/200` request, where the latter is logged at ERROR level:

```
$ make run
mkdir -p dist
CGO_ENABLED=0 go build -ldflags="-s -w" -o dist/go-httpbin ./cmd/go-httpbin
HOST=127.0.0.1 PORT=8080 dist/go-httpbin
time=2024-08-28T14:18:06.611-04:00 level=INFO msg="go-httpbin listening on http://127.0.0.1:8080"
time=2024-08-28T14:18:14.150-04:00 level=ERROR msg="500 GET /status/500 0.0ms" status=500 method=GET uri=/status/500 size_bytes=0 duration_ms=0.01675 user_agent=curl/8.7.1 client_ip=127.0.0.1:60335
time=2024-08-28T14:18:24.516-04:00 level=ERROR msg="200 GET /status/200 0.1ms" status=200 method=GET uri=/status/200 size_bytes=0 duration_ms=0.08220799999999999 user_agent=curl/8.7.1 client_ip=127.0.0.1:60338
```

Reply:
Oh yeah true, sorry about that, will fix it in my next commit.